### PR TITLE
Fix: fix comment handling in padded-blocks when relying on ASI (fixes #2336)

### LINKED ISSUE #2336: Comment at end of block before closing } trigger padded-block 'never' rule.
It appears when you have a comment at the end of a block it is still triggering the `padded-block` rule when set to `'never'`. This is on latest published at time of this issue (v0.19.0). I've included the output, the example, the version command, and our eslintrc below. Let me know if any more information would be helpful.

Command line output:
```
eslint test.js

test.js
  6:1  error  Block must not be padded by blank lines  padded-blocks
```

Example JS (test.js):
```
a()

function a () {
  a()
  // stuff
}
```

eslint version verification:
```
eslint --version
v0.19.0
```

Our .eslintrc:
```
{
  "env": {
    "node": true,
    "browser": false
  },
  "global": [
    "after",
    "afterEach",
    "before",
    "beforeEach",
    "describe",
    "it"
  ],
  "rules": {
    "brace-style": 2,
    "comma-style": [2, "last"],
    "consistent-return": 0,
    "curly": 0,
    "eol-last": 2,
    "eqeqeq": 2,
    "camelcase": 0,
    "dot-notation": 2,
    "func-names": 2,
    "guard-for-in": 2,
    "max-len": [2, 90],
    "max-nested-callbacks": [2, 3],
    "max-params": [2, 5],
    "new-cap": 2,
    "no-comma-dangle": 0,
    "no-console": 1,
    "no-else-return": 2,
    "no-floating-decimal": 2,
    "no-lonely-if": 2,
    "no-mixed-requires": 2,
    "no-multiple-empty-lines": 2,
    "no-new": 2,
    "no-shadow": 1,
    "no-undef": 2,
    "no-underscore-dangle": 0,
    "no-unused-vars": 2,
    "no-use-before-define": 0,
    "padded-blocks": [2, "never"],
    "radix": 2,
    "semi": [2, "never"],
    "space-after-function-name": 0,
    "space-after-keywords": 2,
    "space-before-blocks": 2,
    "space-infix-ops": 2,
    "spaced-line-comment": 2,
    "space-return-throw-case": 2,
    "space-unary-ops": 2,
    "strict": 0,
    "quotes": [0, "single"],
    "wrap-iife": 2
  }
}
```
Past similar issues: https://github.com/eslint/eslint/issues/1938

### PR DESCRIPTION
**Pull request by lo1tuma**

### CODE REVIEW

File: lib/rules/padded-blocks.js [hunk 1]
```diff
@@ -17,6 +17,30 @@ module.exports = function (context) {
         NEVER_MESSAGE = "Block must not be padded by blank lines.";
 
     /**
+     * Retrieves an array of all comments defined inside the given node.
+     * @param {ASTNode} node The AST node.
+     * @returns {ASTNode[]} An array of comment nodes.
+     */
+    function getCommentsInNode(node) {
+        var allComments = context.getAllComments();
```
Comment (nzakas) [inline lib/rules/padded-blocks.js]: This looks like a pretty heavy process to go through for each node that's checked. Is this necessary?

**Comment by nzakas** [comment]
You suspected this might be an Espree bug, did you confirm that?

**Comment by lo1tuma** [comment]
I don’t know exactly how the comment attachment algorithm in espree is supposed to work if relying on ASI.

Example with semicolon:
```js
a(); // comment
```
The trailing comment is attached to the `ExpressionStatement` node.

Example without semicolon:
```js
a() // comment
```
The trailing comment is attached to the `CallExpression` node.

This PR is actually a workaround and does not rely on the correct attachment of trailing comments.

**Comment by nzakas** [comment]
Thanks, I'd like to investigate Espree as the source of the bug of comment attachment is confused by ASI. Otherwise, we will be patching in many places to overcome this difference in behavior. Can you for an Espree bug with this info?

**Comment by btmills** [comment]
Could this be fixed by attaching comments to tokens? That idea was brought up before in [#2211 (comment)](https://github.com/eslint/eslint/issues/2211#issuecomment-91922185).

**Comment by nzakas** [comment]
I'd rather see if it's possible to normalize the comment attachment to nodes. Semicolons really shouldn't affect this so much (I think).

**Comment by lo1tuma** [comment]
Related espree issue eslint/espree/issues/137

**Comment by lo1tuma** [comment]
I’ve updated the PR to also include a fix for #2788.

File: lib/rules/padded-blocks.js (hunk 1 above)
Comment (nzakas) [inline lib/rules/padded-blocks.js]: Still curious if there's a way to optimize this. Starting atvthe first comment each time seems inefficient.

**Pull request merged**